### PR TITLE
[c-str-span] New port

### PR DESCRIPTION
**Pull request by SamuelMarks**
**Describe the pull request**

> UTF-8 replacement for C strings, supporting zero-copy use-cases (non-null-terminated).
> 
> Extracted from https://github.com/Azure/azure-sdk-for-c @ [`ac28b8`](https://github.com/Azure/azure-sdk-for-c/tree/ac28b8af9254798c1e891945e896d2ae2bdf745e) with:
>   - No dependency on other Azure SDK for C functionalities;
>   - C89 (ANSI C) compliance;
>   - CTest, CDash, CPack support.

https://github.com/SamuelMarks/c-str-span [also GitHub Actions added here testing on Windows, Linux, and macOS]

- #### What does your PR fix?
N/A

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?
  <all / linux, windows, ...>, <Yes/No>

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
Yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?
Yes

### CODE REVIEW

**Comment by SamuelMarks** [comment]
Don't merge yet: there's an underscore vs lib vs dash thing going on (try to `find_package` to replicate)

File: ports/c-str-span/vcpkg.json [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  "supports": "!uwp",
+  "dependencies": [
+    {
+      "name": "vcpkg-cmake",
+      "host": true
+    },
+    {
+      "name": "vcpkg-cmake-config",
+      "host": true
+    }
```
Comment (Thomas1664) [inline ports/c-str-span/vcpkg.json:15]:
```suggestion
    }
```
Not needed

**Review by raghavthind2005 — DISMISSED** [review]

**Review by FrankXie05 — DISMISSED** [review]
We recommend adding annotations for better recognition.
````
+ REF f7619819a604b808f7c31b48742336ff26f42061 # v0.0.1
+ "homepage": "https://github.com/SamuelMarks/c-str-span",
````

**Comment by BillyONeal** [comment]
This port causes a usage to be generated like:

```
The package c-str-span provides CMake targets:

    find_package(c_str_span CONFIG REQUIRED)
    target_link_libraries(main PRIVATE c_str_span c_str_span_compiler_flags)
```

It seems like you should be providing a `usage` that says:

```
The package c-str-span provides CMake targets:

    find_package(c_str_span CONFIG REQUIRED)
    target_link_libraries(main PRIVATE c_str_span)
```

(note removed `c_str_span_compiler_flags`)?

Also, since you're also the upstream maintainer, have you considered exporting your CMake targets into a namespace? It leads to better diagnostics for your customers because `c_str_span::c_str_span` must be a CMake target name so CMake won't assume it's a `.lib`, pass it down to the linker, and cause messages from the linker about looking for that.

> Don't merge yet: there's an underscore vs lib vs dash thing going on (try to `find_package` to replicate)

You're currently emitting underscores for your `find_package` name:

image

If you don't want a merge to happen just change the PR into a draft one in the future :)

**Comment by BillyONeal** [comment]
Marking "author response" for the usage question.

**Comment by SamuelMarks** [comment]
@BillyONeal Actually that is an important question, so I take the flags—and most of my CMake structure—from:
https://cmake.org/cmake/help/latest/guide/tutorial/Adding%20Generator%20Expressions.html

Should this not be added to the `install` target?

https://github.com/SamuelMarks/c-str-span/blob/988d819/c_str_span/CMakeLists.txt#L57-L65

**Comment by BillyONeal** [comment]
> @BillyONeal Actually that is an important question, so I take the flags—and most of my CMake structure—from: https://cmake.org/cmake/help/latest/guide/tutorial/Adding%20Generator%20Expressions.html
> 
> Should this not be added to the `install` target?
> 
> https://github.com/SamuelMarks/c-str-span/blob/988d819/c_str_span/CMakeLists.txt#L57-L65

Do you want the user to select that "compiler flags" target individually? I assumed that you want the user to just name c_str_span, and internally have a `target_link_libraries(c_str_span c_str_span_compiler_flags)` so the user doesn't have to name what looks like an internal helper target?

**Comment by BillyONeal** [comment]
Just noting that I did see the pushes but don't think the question here is answered: https://github.com/microsoft/vcpkg/pull/24410#issuecomment-1118056986

**Comment by FrankXie05** [comment]
@SamuelMarks Ping for reply

**Comment by SamuelMarks** [comment]
@BillyONeal How do I hide `c_str_span_compiler_flags` behind the target, but ensure its always included?

**Comment by SamuelMarks** [comment]
@BillyONeal How do I hide `c_str_span_compiler_flags` behind the target, but ensure it's always included?

**Comment by BillyONeal** [comment]
I'm not sure I understand your question. If it's something required by your target `c_str_span`, then `c_str_span` itself should be ensuring things are correct here rather than hoping that people also link with something else. This is kind of the crux of my question above:

> Do you want the user to select that "compiler flags" target individually? I assumed that you want the user to just name c_str_span, and internally have a `target_link_libraries(c_str_span PUBLIC c_str_span_compiler_flags)` so the user doesn't have to name what looks like an internal helper target?

If you want to ensure all callers use those compiler flags, then you should have `target_link_libraries(c_str_span PUBLIC c_str_span_compiler_flags)` which mean linking with `c_str_span` will also bring in the compiler flags. If it's something that's optional for people to use, then the usage you have here is probably correct.

**Comment by FrankXie05** [comment]
Pinging @SamuelMarks  for response. Is work still being done for this PR?

**Comment by SamuelMarks** [comment]
Hmm, do the per target compile flags cause any global effects? - Otherwise I'll do as you suggest and make them `PUBLIC`

**Comment by FrankXie05** [comment]
@SamuelMarks I'm sorry I can't give you an answer
Ping @BillyONeal for reply.

**Comment by BillyONeal** [comment]
> Hmm, do the per target compile flags cause any global effects? - Otherwise I'll do as you suggest and make them `PUBLIC`

They cause effects to anyone who did `target_link_libraries(your-library)`. If you're saying the user must link to both your-library and your-library-compile-flags, just making things public has that effect without the downstream user needing to link with an additional thing.

**Comment by FrankXie05** [comment]
@SamuelMarks Is work still being done for this PR?

**Comment by FrankXie05** [comment]
@SamuelMarks 
For the usage of cmake exported by this port now:
````
The package c-str-span provides CMake targets:

    find_package(c_str_span CONFIG REQUIRED)
    target_link_libraries(main PRIVATE c_str_span c_str_span_compiler_flags)
````
1. Do you really want users to need to link the compilers flags(`c_str_span_compiler_flags`) when using them?
2. If question 1 is no, you can add a usage to tell people that the correct compiler flags should be linked.

**Comment by FrankXie05** [comment]
Pinging @SamuelMarks  for response. Is work still being done for this PR?

**Comment by FrankXie05** [comment]
Draft this PR, since it didn't respond for a long time.

**Comment by FrankXie05** [comment]
Pinging @SamuelMarks  for response. Is work still being done for this PR?

**Comment by SamuelMarks** [comment]
Yes work is still being done here, about to commit up a fix for Windows (MSVC)

**Comment by FrankXie05** [comment]
Closing this PR since it seems that no progress is being made. Please reopen if work is still being done.

**Comment by SamuelMarks** [comment]
@FrankXie05 @BillyONeal Can you reopen this? - Just merged master and updated this port. - Thanks

```
commit 4baa631e7b572149e4aefdcaa3202f310a3190dc (HEAD -> c-str-span)
Author: Samuel Marks <807580+SamuelMarks@users.noreply.github.com>
Date:   Fri Apr 18 11:39:34 2025 -0600

    [c-str-span] 2023-02-11 -> 2025-04-18
```

Thanks to

```
commit 08aff93228c7611beefacbb217c564d87ef69269 (HEAD -> master, origin/master, origin/HEAD)
Author: Samuel Marks <807580+SamuelMarks@users.noreply.github.com>
Date:   Fri Apr 18 11:37:19 2025 -0600

    [c_str_span/c_str_span.h] Implemented `az_span_isalnum`, `az_span_isalpha`,`az_span_isblank`, `az_span_isdigit`, `az_span_islower`, `az_span_isupper` ; [CMakeLists.txt] Fix config to work on newer CMake's
```

**Pull request closed without merging**